File: 345.反转字符串中的元音字母.cs
/*
 * @lc app=leetcode.cn id=345 lang=csharp
 *
 * [345] 反转字符串中的元音字母
 */

// @lc code=start
public class Solution {
    public string ReverseVowels(string s) {
        if(s.Length == 1) return s;
        int l = 0;
        int r = s.Length - 1;
        char[] arr = s.ToCharArray();
        while(l < r)
        {
            if(s[l] != 'a' && s[l] != 'e' && s[l] != 'i' && s[l] != 'o' && s[l] != 'u'
                && s[l] != 'A' && s[l] != 'E' && s[l] != 'I' && s[l] != 'O' && s[l] != 'U')
            {
                l++;
            }
            else if(s[r] != 'a' && s[r] != 'e' && s[r] != 'i' && s[r] != 'o' && s[r] != 'u'
                && s[r] != 'A' && s[r] != 'E' && s[r] != 'I' && s[r] != 'O' && s[r] != 'U')
            {
                r--;
            }
            else if(s[l] == s[r])
            {
                l++;
                r--;
            }
            else
            {
                char c = arr[l];
                arr[l] = arr[r];
                arr[r] = c;
                l++;
                r--;
            }
        }
        return new string(arr);
    }
}
// @lc code=end

File: Answer/345.反转字符串中的元音字母.cpp
/*
 * @lc app=leetcode.cn id=345 lang=cpp
 *
 * [345] 反转字符串中的元音字母
 */

// @lc code=start
class Solution {
public:
    string reverseVowels(string s) {
        if(s.length() < 2) return s;
        char c;
        int i = 0, j = s.length() - 1;
        while(i < j)
        {
            if((tolower(s[i]) == 'a' ||
                tolower(s[i]) == 'e' ||
                tolower(s[i]) == 'i' ||
                tolower(s[i]) == 'o' ||
                tolower(s[i]) == 'u') &&
               (tolower(s[j]) == 'a' ||
                tolower(s[j]) == 'e' ||
                tolower(s[j]) == 'i' ||
                tolower(s[j]) == 'o' ||
                tolower(s[j]) == 'u'))
            {
                c = s[i];
                s[i] = s[j];
                s[j] = c;
                i++;
                j--;
            }
            else if(tolower(s[i]) == 'a' ||
                    tolower(s[i]) == 'e' ||
                    tolower(s[i]) == 'i' ||
                    tolower(s[i]) == 'o' ||
                    tolower(s[i]) == 'u')
            {
                j--;
            }
            else if(tolower(s[j]) == 'a' ||
                    tolower(s[j]) == 'e' ||
                    tolower(s[j]) == 'i' ||
                    tolower(s[j]) == 'o' ||
                    tolower(s[j]) == 'u')
            {
                i++;
            }
            else
            {
                i++;
                j--;
            }
        }
        return s;
    }
};
// @lc code=end

